CORS-2899: AWS/CAPI Local Zones network provisioning

mtulio opened the pull request:
[CORS-2899](https://issues.redhat.com/browse/CORS-2899)

This PR introduce support of AWS Local Zones using CAPA as infrastructure provisioner for managed (automated VPC) and unmanaged (BYO VPC) deployments.

References:
- Jira story on "CAPA moving to Terraform" Epic/Goal: https://issues.redhat.com/browse/CORS-2899
- CAPA feature request: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4874
- CAPA feature PR: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882

Blocked by:
- [x] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882
- [x] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4948

mtulio commented:
/test ?

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
Two success cluster creation on Local Zones. Triggering other jobs.

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
The current version is provisioning the cluster, tested both local and CI: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/8173/pull-ci-openshift-installer-master-altinfra-e2e-aws-ovn-localzones/1770083285305659392

![Screenshot from 2024-03-19 14-02-05](https://github.com/openshift/installer/assets/3216894/027f4b87-30d6-49bc-ae38-e486dc5e3bc5)

Local tests, without specifying the zones for Control Plane, worker or defaultPlatform:

```
$ oc get nodes
NAME                          STATUS   ROLES                  AGE    VERSION
ip-10-0-20-247.ec2.internal   Ready    control-plane,master   172m   v1.29.1+5a4819c
ip-10-0-23-113.ec2.internal   Ready    worker                 163m   v1.29.1+5a4819c
ip-10-0-35-216.ec2.internal   Ready    worker                 163m   v1.29.1+5a4819c
ip-10-0-39-43.ec2.internal    Ready    control-plane,master   173m   v1.29.1+5a4819c
ip-10-0-52-70.ec2.internal    Ready    control-plane,master   173m   v1.29.1+5a4819c
ip-10-0-57-99.ec2.internal    Ready    worker                 163m   v1.29.1+5a4819c
ip-10-0-6-55.ec2.internal     Ready    edge,worker            159m   v1.29.1+5a4819c

$ oc get machineset -A
NAMESPACE               NAME                                        DESIRED   CURRENT   READY   AVAILABLE   AGE
openshift-machine-api   mrb-capalz-01-qqfc5-edge-us-east-1-nyc-1a   1         1         1       1           173m
openshift-machine-api   mrb-capalz-01-qqfc5-worker-us-east-1a       1         1         1       1           173m
openshift-machine-api   mrb-capalz-01-qqfc5-worker-us-east-1b       1         1         1       1           173m
openshift-machine-api   mrb-capalz-01-qqfc5-worker-us-east-1c       1         1         1       1           173m
openshift-machine-api   mrb-capalz-01-qqfc5-worker-us-east-1d       0         0                             173m
openshift-machine-api   mrb-capalz-01-qqfc5-worker-us-east-1f       0         0                             173m

```

mtulio commented:
/cc @r4f4 @patrickdillon 
/uncc rna-afk pawanpinjarkar

mtulio commented:
I just provided some improvements in the subnet CIDR blocks handling when edge zones VS other (default zones).
Edge subnets uses a lower number of blocks, consuming from one slot from the regular CIDR Pool (sliced from VPC CIDR).

To prevent IP exhaustion for public blocks for edge subnets, I am preserving the logic of slicing 50% of edge CIDR pool for public subnets in edge zones, same behavior of terraform, instead of slice (again) smaller blocks for public edge subnets.

Deploying a cluster in many Local Zones, with VPC CIDR /16, we would be able to see something like this:

```
$ oc get nodes
NAME                           STATUS   ROLES                  AGE     VERSION
ip-10-0-11-75.ec2.internal     Ready    control-plane,master   42m     v1.29.1+5a4819c
ip-10-0-112-46.ec2.internal    Ready    edge,worker            26m     v1.29.1+5a4819c
ip-10-0-113-42.ec2.internal    Ready    edge,worker            27m     v1.29.1+5a4819c
ip-10-0-114-113.ec2.internal   Ready    edge,worker            6m54s   v1.29.1+5a4819c
ip-10-0-115-29.ec2.internal    Ready    edge,worker            25m     v1.29.1+5a4819c
ip-10-0-116-83.ec2.internal    Ready    edge,worker            24m     v1.29.1+5a4819c
ip-10-0-117-73.ec2.internal    Ready    edge,worker            24m     v1.29.1+5a4819c
ip-10-0-118-97.ec2.internal    Ready    edge,worker            14m     v1.29.1+5a4819c
ip-10-0-119-24.ec2.internal    Ready    edge,worker            24m     v1.29.1+5a4819c
ip-10-0-120-77.ec2.internal    Ready    edge,worker            24m     v1.29.1+5a4819c
ip-10-0-121-28.ec2.internal    Ready    edge,worker            22m     v1.29.1+5a4819c
ip-10-0-122-52.ec2.internal    Ready    edge,worker            29m     v1.29.1+5a4819c
ip-10-0-123-62.ec2.internal    Ready    edge,worker            29m     v1.29.1+5a4819c
ip-10-0-124-60.ec2.internal    Ready    edge,worker            20m     v1.29.1+5a4819c
ip-10-0-125-44.ec2.internal    Ready    edge,worker            13m     v1.29.1+5a4819c
ip-10-0-22-109.ec2.internal    Ready    control-plane,master   41m     v1.29.1+5a4819c
ip-10-0-30-104.ec2.internal    Ready    worker                 32m     v1.29.1+5a4819c
ip-10-0-37-214.ec2.internal    Ready    control-plane,master   42m     v1.29.1+5a4819c
ip-10-0-41-212.ec2.internal    Ready    worker                 32m     v1.29.1+5a4819c
ip-10-0-8-127.ec2.internal     Ready    worker                 32m     v1.29.1+5a4819c

$ oc get machines -A
NAMESPACE               NAME                                              PHASE     TYPE          REGION      ZONE               AGE
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-atl-1a-9tvx2   Running   c5d.2xlarge   us-east-1   us-east-1-atl-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-bos-1a-rfs6s   Running   m5.xlarge     us-east-1   us-east-1-bos-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-bue-1a-ptbc2   Running   c5.2xlarge    us-east-1   us-east-1-bue-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-chi-1a-lstvn   Running   c5d.2xlarge   us-east-1   us-east-1-chi-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-dfw-1a-28wmm   Running   c5d.2xlarge   us-east-1   us-east-1-dfw-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-iah-1a-scjlx   Running   c5d.2xlarge   us-east-1   us-east-1-iah-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-lim-1a-9nfr6   Running   r5.xlarge     us-east-1   us-east-1-lim-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-mci-1a-28cxn   Running   c5d.2xlarge   us-east-1   us-east-1-mci-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-mia-1a-85qjt   Running   m5.xlarge     us-east-1   us-east-1-mia-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-msp-1a-m9m5p   Running   c5d.2xlarge   us-east-1   us-east-1-msp-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-nyc-1a-rmzhr   Running   c5d.2xlarge   us-east-1   us-east-1-nyc-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-phl-1a-b9766   Running   c5d.2xlarge   us-east-1   us-east-1-phl-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-qro-1a-zscrz   Running   c5.2xlarge    us-east-1   us-east-1-qro-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-edge-us-east-1-scl-1a-5wrgb   Running   c5.2xlarge    us-east-1   us-east-1-scl-1a   38m
openshift-machine-api   mrb-capalz-04-shcgp-master-0                      Running   m6i.xlarge    us-east-1   us-east-1a         42m
openshift-machine-api   mrb-capalz-04-shcgp-master-1                      Running   m6i.xlarge    us-east-1   us-east-1b         42m
openshift-machine-api   mrb-capalz-04-shcgp-master-2                      Running   m6i.xlarge    us-east-1   us-east-1c         42m
openshift-machine-api   mrb-capalz-04-shcgp-worker-us-east-1a-xzbs9       Running   m6i.xlarge    us-east-1   us-east-1a         38m
openshift-machine-api   mrb-capalz-04-shcgp-worker-us-east-1b-89rbc       Running   m6i.xlarge    us-east-1   us-east-1b         38m
openshift-machine-api   mrb-capalz-04-shcgp-worker-us-east-1c-cpvn6       Running   m6i.xlarge    us-east-1   us-east-1c         38m

```

![Screenshot from 2024-03-19 17-48-47](https://github.com/openshift/installer/assets/3216894/f9ac7160-a8ab-4f4d-b92b-0a2f301f6f98)

![Screenshot from 2024-03-19 17-48-54](https://github.com/openshift/installer/assets/3216894/8d03adda-64c4-493d-b26c-e67c80de43c8)

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

/test altinfra-e2e-aws-ovn-shared-vpc-edge-zones

mtulio commented:
/test unit

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test ?

mtulio commented:
/test altinfra-e2e-aws-ovn-shared-vpc-local-zones

mtulio commented:
/test altinfra-e2e-aws-ovn

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test altinfra-e2e-aws-ovn-shared-vpc-local-zones

mtulio commented:
/test altinfra-e2e-aws-ovn

mtulio commented:
/test unit

mtulio commented:
This PR is blocked by https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4948
/hold

mtulio commented on pkg/asset/manifests/aws/zones_test.go:
```diff
@@ -18,7 +19,9 @@ import (
 )
 
-var stubDefaultCIDR = "10.0.0.0/16"
+var (
+	stubDefaultCIDR = "10.0.0.0/16"
+)
```
var stubDefaultCIDR = "10.0.0.0/16"

mtulio commented on pkg/asset/manifests/aws/zones_test.go:
```diff
@@ -224,78 +256,163 @@ func Test_setSubnetsManagedVPC(t *testing.T) {
-				if !tt.wantErr {
-					t.Errorf("setSubnetsManagedVPC() #2 error: %v, wantErr: %v", err, tt.wantErr)
-				}
+			// TODO review those empty test cases
```
still TODO?

mtulio commented on pkg/asset/manifests/aws/zones_test.go:
```diff
@@ -224,78 +256,163 @@ func Test_setSubnetsManagedVPC(t *testing.T) {
-			got.Subnets = tSortCapaSubnetsByID(got.Subnets)
-			if tt.want != nil {
-				assert.EqualValuesf(t, tt.want, got, "%v failed.\nWant: %+v\nGot: %+v", tt.name)
+			tt.args.in.Cluster.Spec.NetworkSpec.Subnets = tSortCapaSubnetsByID(tt.args.in.Cluster.Spec.NetworkSpec.Subnets)
+			if got := tt.args.in.Cluster.Spec.NetworkSpec; !cmp.Equal(got, tt.want) {
+				t.Errorf("setSubnetsManagedVPC() NetworkSpec.Subnets diff: %v", cmp.Diff(got, tt.want))
```
@r4f4 is there a better alternative available on installer of `cmp` providing a friendly `Diff` of complex structures?

mtulio replied:
nvm, managed it.

mtulio reviewed (commented):
nit's on units for edge zones

mtulio commented:
/test golint

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones

mtulio commented:
/test golint

mtulio commented:
> /test altinfra-e2e-aws-ovn-localzones

Managed VPC finished successfully. I can see machine created:

> Note: best-effort instance type also chose current generation (`m6i.xlarge`) on supported (newer) Local Zone location (`us-west-2-phx-2a`)

```sh
$ omc get machines.machine.openshift.io -n openshift-machine-api
NAME                                                     PHASE     TYPE         REGION      ZONE               AGE   NODE                                         PROVIDERID                                    STATE
ci-op-c3x11ysy-c595f-8crzq-edge-us-west-2-phx-2a-l2wr4   Running   m6i.xlarge   us-west-2   us-west-2-phx-2a   3h    ip-10-0-101-102.us-west-2.compute.internal   aws:///us-west-2-phx-2a/i-064df93bf5d80edb5   running
ci-op-c3x11ysy-c595f-8crzq-master-0                      Running   m6a.xlarge   us-west-2   us-west-2a         3h    ip-10-0-0-169.us-west-2.compute.internal     aws:///us-west-2a/i-0d5e9bb6d458024ec         running
ci-op-c3x11ysy-c595f-8crzq-master-1                      Running   m6a.xlarge   us-west-2   us-west-2a         3h    ip-10-0-19-239.us-west-2.compute.internal    aws:///us-west-2a/i-0a3621df2423b9880         running
ci-op-c3x11ysy-c595f-8crzq-master-2                      Running   m6a.xlarge   us-west-2   us-west-2a         3h    ip-10-0-30-3.us-west-2.compute.internal      aws:///us-west-2a/i-06608c254560477cd         running
ci-op-c3x11ysy-c595f-8crzq-worker-us-west-2a-jc9kz       Running   m6a.xlarge   us-west-2   us-west-2a         3h    ip-10-0-8-117.us-west-2.compute.internal     aws:///us-west-2a/i-0659d6d0ea8753ae8         running
ci-op-c3x11ysy-c595f-8crzq-worker-us-west-2a-lk2hs       Running   m6a.xlarge   us-west-2   us-west-2a         3h    ip-10-0-10-217.us-west-2.compute.internal    aws:///us-west-2a/i-0fb741faeae40be47         running
ci-op-c3x11ysy-c595f-8crzq-worker-us-west-2b-5mzf4       Running   m6a.xlarge   us-west-2   us-west-2b         3h    ip-10-0-37-237.us-west-2.compute.internal    aws:///us-west-2b/i-02bfbe4a9ade65d2c         running

```

Testing BYO VPC (unmanaged):

/test altinfra-e2e-aws-ovn-shared-vpc-local-zones

mtulio commented:
/test unit

mtulio commented:
PR rebased after https://github.com/openshift/installer/pull/8223 and local vendoring removed.

Running the feature tests:
/test altinfra-e2e-aws-ovn-localzones
/test altinfra-e2e-aws-ovn-shared-vpc-local-zones

r4f4 commented on pkg/asset/manifests/aws/zones.go:
```diff
@@ -49,6 +50,9 @@ func (zin *zonesInput) GatherSubnetsFromMetadata(ctx context.Context) (err error
 		return fmt.Errorf("failed to get public subnets: %w", err)
 	}
+	if zin.Subnets.edgeSubnets, err = zin.InstallConfig.AWS.EdgeSubnets(ctx); err != nil {
+		return fmt.Errorf("failed to get public subnets: %w", err)
```
```suggestion
		return fmt.Errorf("failed to get edge subnets: %w", err)
```

r4f4 commented on pkg/asset/manifests/aws/zones.go:
```diff
@@ -178,45 +200,60 @@ func setSubnetsManagedVPC(in *zonesInput) error {
+	publicCIDR := privateCIDRs[len(allAvailabilityZones)].String()
+	var edgeCIDR string
+	if len(allEdgeZones) > 0 {
+		publicCIDR = privateCIDRs[len(allAvailabilityZones)].String()
```
```suggestion

```
No need to repeat this line.

r4f4 commented on pkg/asset/manifests/aws/zones.go:
```diff
@@ -232,6 +269,50 @@ func setSubnetsManagedVPC(in *zonesInput) error {
 	}
+
+	// Create subnets from edge zone pool with type local-zone.
+	if len(allEdgeZones) > 0 {
```
What about we invert the check here and return early? That would decrease the indent level by 1.
```suggestion
    // no edge zones, nothing else to do
	if len(allEdgeZones) == 0 {
	    return nil
	}
```

r4f4 commented on pkg/asset/manifests/aws/zones.go:
```diff
@@ -263,6 +345,7 @@ func extractZonesFromInstallConfig(in *zonesInput) (*zonesCAPI, error) {
 		// See https://github.com/openshift/installer/pull/8173
 		if pool.Name == types.MachinePoolEdgeRoleName {
+			out.edgeZones.Insert(pool.Platform.AWS.Zones...)
 			continue
```
Do you still want this continue here?

r4f4 replied:
Ah, if it's an edge pool you don't want to `SetDefaultConfigZones`.

mtulio replied:
that's correct, edge zones arent eligible for regular zones

mtulio replied:
edge zones isn't eligible to set default zones (next step), so continue here is correct. I believe it is missing a check for some unexpected situation when there is no zones defined in the edge pool (or should we skip this considering the [validation is performing](https://github.com/openshift/installer/blob/1c9f90d42d2869a6fe95848b9e1c8ee9cda401d4/pkg/asset/installconfig/aws/validation.go#L264-L276) those checks?).

I added it in my recent changes just to make sure a coverage in the tests.

r4f4 commented on pkg/asset/manifests/aws/zones.go:
```diff
@@ -263,6 +345,7 @@ func extractZonesFromInstallConfig(in *zonesInput) (*zonesCAPI, error) {
 		// Ignoring as edge pool is not yet supported by CAPA.
 		// See https://github.com/openshift/installer/pull/8173
```
stale comment?

mtulio commented:
Converting to draft to address the review feedback, and adding more scenarios in the unit tests for cluster CAPA zones.

mtulio commented:
/test unit
/test golint

mtulio commented:
@r4f4 thanks for your review.

I found one "issue" (mentioned on Slack) which was specifically [here](https://github.com/openshift/installer/compare/1c9f90d42d2869a6fe95848b9e1c8ee9cda401d4..854f286c2bdb6ce772da027e3d8e648dafdcb7c6#diff-343fc6eb1760a5cba9040fe31edb8be6aabb12a355e9615c3b29156894c3ad98R371-R374) (when the worker pool isn't defined, the zone wasn't added). In the practice, it will not generate any issue for the provisioner as the default zones are set earlier, instead a issue caught in the unit validation. I am checking if we'll need to simplify/drop some validations considering the goal is only discover which zones should be used to create subnets, regardless it was added from pools, defaults, or metadata.

The [ZonesCAPI](https://github.com/openshift/installer/compare/1c9f90d42d2869a6fe95848b9e1c8ee9cda401d4..854f286c2bdb6ce772da027e3d8e648dafdcb7c6#diff-343fc6eb1760a5cba9040fe31edb8be6aabb12a355e9615c3b29156894c3ad98R65-R71) had been exported and generated a lot of changes in the whole zone code in favor of getting better unit tests feedback using go-cmp.

The remaining changes are addressing your feedback.

mtulio commented:
/test altinfra-e2e-aws-ovn-localzones
/test altinfra-e2e-aws-ovn-shared-vpc-local-zones

mtulio commented:
/uncc rna-afk sadasu barbacbd
/assign @r4f4 @patrickdillon

r4f4 reviewed (commented):
/lgtm

mtulio commented:
/hold cancel

patrickdillon commented:
/approve
